Pointwise log-likelihoods for non-factorizable models

marcobonici opened the pull request:
Closes #49 .

sethaxen commented on src/loo_nonfactorizable.jl:
```diff
@@ -0,0 +1,68 @@
+whitened_residuals(F::Cholesky, r::AbstractVector) = F \ r
+whitened_residuals(F::AbstractMatrix, r::AbstractVector) = F * r
+
+function pointwise_normal_loglikelihood!(
```
I prefer to drop the method that takes mean and cov (or `Cholesky`) directly and just keep the one that takes an `MvNormal` or `MvNormalCanon`.

marcobonici replied:
Agreed, comment included.

sethaxen commented on src/loo_nonfactorizable.jl:
```diff
@@ -0,0 +1,68 @@
+    return nothing
+end
+
+function elements_from_d(d::Distributions.MvNormal)
+    μ = d.μ
+    F = d.Σ.chol
+    c = diag_cov_from_cov_chol(F)
+    return μ, F, c
+end
+
+function elements_from_d(d::Distributions.MvNormalCanon)
+    μ = d.μ
+    iΣ = d.J
+    c = diag(iΣ)
+    return μ, iΣ, c
+end
```
These names aren't very informative, but more importantly, I think it's not necessary for these two to have this utility method at all, e.g. here's an implementation I had come up with:
```julia
function pointwise_loglikelihood!(logl::AbstractVector{<:Real}, dist::MvNormal, y::AbstractVector{<:Real})
    Σ = cov(dist)
    λ = _pd_diag_inv(Σ)
    g = Σ \ (y - mean(dist))
    return @. logl = (log(λ) - g^2 / λ - log2π) / 2
end
function pointwise_loglikelihood!(logl::AbstractVector{<:Real}, dist::MvNormalCanon, y::AbstractVector{<:Real})
    J = invcov(dist)
    λ = diag(J)
    cov_inv_y = J * y
    return @. logl = (log(λ) - (cov_inv_y - dist.h)^2 / λ - log2π) / 2
end
```

sethaxen commented on src/loo_nonfactorizable.jl:
```diff
@@ -0,0 +1,68 @@
+    return μ, iΣ, c
+end
+
+function pointwise_conditional_loglikelihood(
```
Let's drop the `conditional`; it's a mouthful.

sethaxen commented on src/loo_nonfactorizable.jl:
```diff
@@ -0,0 +1,68 @@
+    return logl
+end
+
+function loo_nonfactorized(
```
Since this is such a thin wrapper, I prefer to leave out this method to keep the API minimal. Rather, we'll eventually include a docs page that explains how to use the utility method above with factorized models.

marcobonici replied:
I have removed it as suggested.

sethaxen commented on src/loo_nonfactorizable.jl:
Below I recommend renaming `pointwise_conditional_loglikelihood` to `pointwise_loglikelihoods` and dropping `loo_nonfactorized`, so perhaps rename to `pointwise_loglikelihoods.jl`? Another benefit of this is that Turing has need of a utility function for computing pointwise log-likelihoods (currently it uses one defined in ParetoSmooth, but as it moves to PosteriorStats/PSIS, this would need to move to Turing/DynamicPPL, see TuringLang/Turing.jl#2534). In anticipation of other Bayesian packages needing such a utility, we could actually just define `pointwise_loglikelihoods` as a utility function we export that any package could overload as they see fit to support LOO, which would reduce the chances of namespace clashes.

sethaxen reviewed (changes requested):
Thanks for the PR! Overall this looks great! I have a few suggestions.

sethaxen commented on src/loo_nonfactorizable.jl:
```diff
@@ -0,0 +1,68 @@
+    end
+end
+
+whitened_residuals(F::Cholesky, r::AbstractVector) = F \ r
+whitened_residuals(F::AbstractMatrix, r::AbstractVector) = F * r
```
I think it's better to drop all dependence on `Cholesky` (see below), and I don't think we'll sacrifice any efficiency in doing so.

marcobonici commented on src/loo_nonfactorizable.jl:
```diff
@@ -0,0 +1,68 @@
+function diag_cov_from_cov_chol(F::Cholesky)
```
Agreed, I have included your comments.

sethaxen commented on Project.toml:
```diff
@@ -51,6 +52,7 @@ Markdown = "1.6"
 OffsetArrays = "1.1.3"
 Optim = "1.7.2"
 OrderedCollections = "1.3.0"
+PDMats = "0.11.35"
```
It'll be easier to check when the test suite is running (the downgrade jobs will check this), but I think we can relax these bounds to those of our Distributions lower bound:
```suggestion
PDMats = "0.10, 0.11"
```

sethaxen commented on src/PosteriorStats.jl:
```diff
@@ -2,20 +2,31 @@ module PosteriorStats
 using Compat: @constprop
 using DataInterpolations: DataInterpolations
-using Distributions: Distributions
+using Distributions: cov, Distributions, invcov, mean
```
For clarity, I prefer to (in most cases) prefix functions from other packages with the package name:
```suggestion
using Distributions: Distributions
```

sethaxen commented on src/PosteriorStats.jl:
```diff
@@ -2,20 +2,31 @@ module PosteriorStats
 using IteratorInterfaceExtensions: IteratorInterfaceExtensions
 using KernelDensity: KernelDensity
-using LinearAlgebra: mul!, norm, normalize
+using LinearAlgebra:
+    cholesky,
+    Cholesky,
+    diag,
+    Diagonal,
+    LowerTriangular,
+    mul!,
+    norm,
+    normalize,
+    Symmetric,
+    UpperTriangular
```
Only `diag` and `Diagonal` are used, and each only once, so maybe let's not import these either:
```suggestion
using LinearAlgebra: mul!, norm, normalize
```

sethaxen commented on src/PosteriorStats.jl:
```diff
@@ -58,6 +69,7 @@ include("eti.jl")
 include("hdi.jl")
 include("elpdresult.jl")
 include("loo.jl")
+include("pointwise_loglikelihoods.jl")
```
```suggestion
include("pointwise_loglikelihoods.jl")
include("loo.jl")
```

sethaxen commented on src/pointwise_loglikelihoods.jl:
Minor point, but I prefer to define the API functions at the top and define the internal functions below, so can you reorder them to:
```julia
pointwise_loglikelihoods
pointwise_loglikelihoods!
_pd_diag_inv
```

marcobonici replied:
Done!

sethaxen commented on src/pointwise_loglikelihoods.jl:
```diff
@@ -0,0 +1,40 @@
+function _pd_diag_inv(A::PDMats.AbstractPDMat)
+    T = typeof(float(oneunit(eltype(A))))
+    I = Diagonal(ones(T, axes(A, 1)))
```
```suggestion
    I = LinearAlgebra.Diagonal(ones(T, axes(A, 1)))
```

sethaxen commented on src/pointwise_loglikelihoods.jl:
```diff
@@ -0,0 +1,40 @@
+function pointwise_loglikelihood!(
+    logl::AbstractVector{<:Real}, y::AbstractVector{<:Real}, dist::Distributions.MvNormal
+)
+    Σ = cov(dist)
```
```suggestion
    Σ = Distributions.cov(dist)
```

sethaxen commented on src/pointwise_loglikelihoods.jl:
```diff
@@ -0,0 +1,40 @@
+)
+    Σ = cov(dist)
+    λ = _pd_diag_inv(Σ)
+    g = Σ \ (y - mean(dist))
```
```suggestion
    g = Σ \ (y - Distributions.mean(dist))
```

sethaxen commented on src/pointwise_loglikelihoods.jl:
```diff
@@ -0,0 +1,40 @@
+    y::AbstractVector{<:Real},
+    dist::Distributions.MvNormalCanon,
+)
+    J = invcov(dist)
```
```suggestion
    J = Distributions.invcov(dist)
```

sethaxen commented on src/pointwise_loglikelihoods.jl:
```diff
@@ -0,0 +1,40 @@
+    dist::Distributions.MvNormalCanon,
+)
+    J = invcov(dist)
+    λ = diag(J)
```
```suggestion
    λ = LinearAlgebra.diag(J)
```

sethaxen commented on src/pointwise_loglikelihoods.jl:
```diff
@@ -0,0 +1,40 @@
+end
+
+function pointwise_loglikelihood!(
+    logl::AbstractVector{<:Real}, y::AbstractVector{<:Real}, dist::Distributions.MvNormal
```
For clarity, let's rename `logl` to `log_like` for consistency with our `loo` implementation:
```suggestion
    log_like::AbstractVector{<:Real},
    y::AbstractVector{<:Real},
    dist::Distributions.MvNormal,
```

sethaxen commented on src/pointwise_loglikelihoods.jl:
```diff
@@ -0,0 +1,40 @@
+    Σ = cov(dist)
+    λ = _pd_diag_inv(Σ)
+    g = Σ \ (y - mean(dist))
+    return @. logl = (log(λ) - g^2 / λ - log2π) / 2
```
```suggestion
    return @. log_like = (log(λ) - g^2 / λ - log2π) / 2
```

sethaxen commented on src/pointwise_loglikelihoods.jl:
```diff
@@ -0,0 +1,40 @@
+end
+
+function pointwise_loglikelihood!(
+    logl::AbstractVector{<:Real},
```
```suggestion
    log_like::AbstractVector{<:Real},
```

sethaxen commented on src/pointwise_loglikelihoods.jl:
```diff
@@ -0,0 +1,40 @@
+    J = invcov(dist)
+    λ = diag(J)
+    cov_inv_y = J * y
+    return @. logl = (log(λ) - (cov_inv_y - dist.h)^2 / λ - log2π) / 2
```
```suggestion
    return @. log_like = (log(λ) - (cov_inv_y - dist.h)^2 / λ - log2π) / 2
```

sethaxen commented on src/pointwise_loglikelihoods.jl:
```diff
@@ -0,0 +1,40 @@
+    return PDMats.invquad(A, I)
+end
+
+function pointwise_loglikelihood!(
```
Since it's still computing multiple likelihoods, maybe the plural name makes more sense:
```suggestion
function pointwise_loglikelihoods!(
```

sethaxen commented on src/pointwise_loglikelihoods.jl:
```diff
@@ -0,0 +1,40 @@
+    return @. logl = (log(λ) - g^2 / λ - log2π) / 2
+end
+
+function pointwise_loglikelihood!(
```
```suggestion
function pointwise_loglikelihoods!(
```

sethaxen commented on src/pointwise_loglikelihoods.jl:
```diff
@@ -0,0 +1,40 @@
+end
+
+function pointwise_loglikelihoods(
+    obs::AbstractArray{<:Real,N},
```
For consistency with our `loo_pit` signature:
```suggestion
    y::AbstractArray{<:Real,N},
```

sethaxen commented on src/pointwise_loglikelihoods.jl:
```diff
@@ -0,0 +1,40 @@
+        <:Distributions.Distribution{<:Distributions.ArrayLikeVariate{N}},M
+    },
+) where {M,N}
+    T = typeof(log(one(promote_type(eltype(obs), eltype(eltype(dists))))))
```
```suggestion
    T = typeof(log(one(promote_type(eltype(y), eltype(eltype(dists))))))
```

sethaxen commented on src/pointwise_loglikelihoods.jl:
```diff
@@ -0,0 +1,40 @@
+) where {M,N}
+    T = typeof(log(one(promote_type(eltype(obs), eltype(eltype(dists))))))
+    sample_dims = ntuple(identity, M)
+    logl = similar(obs, T, (axes(dists)..., axes(obs)...))
```
```suggestion
    log_like = similar(y, T, (axes(dists)..., axes(y)...))
```

sethaxen commented on src/pointwise_loglikelihoods.jl:
```diff
@@ -0,0 +1,40 @@
+    T = typeof(log(one(promote_type(eltype(obs), eltype(eltype(dists))))))
+    sample_dims = ntuple(identity, M)
+    logl = similar(obs, T, (axes(dists)..., axes(obs)...))
+    for (dist, logl_i) in zip(dists, eachslice(logl; dims=sample_dims))
```
```suggestion
    for (dist, ll) in zip(dists, eachslice(log_like; dims=sample_dims))
```

sethaxen commented on src/pointwise_loglikelihoods.jl:
```diff
@@ -0,0 +1,40 @@
+    sample_dims = ntuple(identity, M)
+    logl = similar(obs, T, (axes(dists)..., axes(obs)...))
+    for (dist, logl_i) in zip(dists, eachslice(logl; dims=sample_dims))
+        pointwise_loglikelihood!(logl_i, obs, dist)
```
```suggestion
        pointwise_loglikelihood!(ll, y, dist)
```

sethaxen commented on src/pointwise_loglikelihoods.jl:
```diff
@@ -0,0 +1,40 @@
+    for (dist, logl_i) in zip(dists, eachslice(logl; dims=sample_dims))
+        pointwise_loglikelihood!(logl_i, obs, dist)
+    end
+    return logl
```
```suggestion
    return log_like
```

sethaxen reviewed (changes requested):
> Hi @sethaxen , I should have addressed your comments (which basically means I have included your suggestions).
> 
> Let me know if any other modification to the PR is needed. Once you are happy with it, I will work on the unit tests and docs.

Perfect, thanks! Here's the final round of minor requests. After this, I think only tests and docstrings remain.

Please include a test that if `y` and `dists` both are `OffsetArray`s then the output is an `OffsetArray` with the matching axes (no need to make the parameters of the `dist` also `OffsetArray`s). You can base this off of the test in https://github.com/arviz-devs/PosteriorStats.jl/blob/main/test/loo.jl

marcobonici commented:
I have some troubles with the `OffsetArray`s stuff.

If I have this code, it works nicely
```julia
Random.seed!(1234)
y = OffsetArray(randn(3), 1:3)

μ = zeros(3)
Σ = Matrix(I, 3, 3)
d = MvNormal(μ, Σ)

# Build y as an OffsetArray of 3-vectors with the same axes as dists
ax = -2:1  # any non-1-based range works

# dists: OffsetVector of MvNormals
dists = OffsetArray(fill(d, length(ax)), ax)

pll = PosteriorStats.pointwise_loglikelihoods(y, dists)

4×3 OffsetArray(::Matrix{Float64}, -2:1, 1:3) with eltype Float64 with indices -2:1×1:3:
 -0.983641  -1.50995  -1.00697
 -0.983641  -1.50995  -1.00697
 -0.983641  -1.50995  -1.00697
 -0.983641  -1.50995  -1.00697
```
It looks to me the axes are preserved as expected.
But if I change to a slightly more interesting case `y = OffsetArray(randn(3), 0:2)` I get
```julia
DimensionMismatch: a has axes (0:2,), b has axes (1:3,), mismatch at dim 1

Stacktrace:
 [1] throw_promote_shape_mismatch(a::Tuple{OffsetArrays.IdOffsetRange{Int64, Base.OneTo{Int64}}}, b::Tuple{OffsetArrays.IdOffsetRange{Int64, Base.OneTo{Int64}}}, i::Int64)
   @ Base ./indices.jl:135
 [2] promote_shape
   @ ./indices.jl:196 [inlined]
 [3] promote_shape
   @ ./indices.jl:188 [inlined]
 [4] -(A::OffsetVector{Float64, Vector{Float64}}, B::OffsetVector{Float64, Vector{Float64}})
   @ Base ./arraymath.jl:7
 [5] pointwise_loglikelihoods!(log_like::SubArray{Float64, 1, OffsetMatrix{Float64, Matrix{Float64}}, Tuple{Int64, Base.Slice{OffsetArrays.IdOffsetRange{Int64, Base.OneTo{Int64}}}}, true}, y::OffsetVector{Float64, Vector{Float64}}, dist::MvNormal{Float64, PDMat{Float64, Matrix{Float64}}, OffsetVector{Float64, Vector{Float64}}})
   @ PosteriorStats [~/Desktop/work/PosteriorStats.jl/src/pointwise_loglikelihoods.jl:23](http://localhost:8888/home/mbonici/Desktop/work/PosteriorStats.jl/src/pointwise_loglikelihoods.jl#line=22)
 [6] pointwise_loglikelihoods(y::OffsetVector{Float64, Vector{Float64}}, dists::OffsetVector{MvNormal{Float64, PDMat{Float64, Matrix{Float64}}, OffsetVector{Float64, Vector{Float64}}}, Vector{MvNormal{Float64, PDMat{Float64, Matrix{Float64}}, OffsetVector{Float64, Vector{Float64}}}}})
   @ PosteriorStats [~/Desktop/work/PosteriorStats.jl/src/pointwise_loglikelihoods.jl:11](http://localhost:8888/home/mbonici/Desktop/work/PosteriorStats.jl/src/pointwise_loglikelihoods.jl#line=10)
 [7] top-level scope
   @ In[141]:14
   ```
   
  Mmmhh. If I change `μ = OffsetArray(zeros(3), 0:2)` I get

```julia
ArgumentError: offset arrays are not supported but got an array with index other than 1

Stacktrace:
 [1] require_one_based_indexing
   @ ./abstractarray.jl:131 [inlined]
 [2] ldiv(F::Cholesky{Float64, Matrix{Float64}}, B::OffsetVector{Float64, Vector{Float64}})
   @ LinearAlgebra [~/.julia/juliaup/julia-1.11.5](http://localhost:8888/home/mbonici/.julia/juliaup/julia-1.11.5)+0.x64.linux.gnu[/share/julia/stdlib/v1.11/LinearAlgebra/src/LinearAlgebra.jl:636](http://localhost:8888/share/julia/stdlib/v1.11/LinearAlgebra/src/LinearAlgebra.jl#line=635)
 [3] \
   @ [~/.julia/juliaup/julia-1.11.5](http://localhost:8888/home/mbonici/.julia/juliaup/julia-1.11.5)+0.x64.linux.gnu[/share/julia/stdlib/v1.11/LinearAlgebra/src/LinearAlgebra.jl:631](http://localhost:8888/share/julia/stdlib/v1.11/LinearAlgebra/src/LinearAlgebra.jl#line=630) [inlined]
 [4] \
   @ [~/.julia/packages/PDMats/cb4LS/src/pdmat.jl:99](http://localhost:8888/home/mbonici/.julia/packages/PDMats/cb4LS/src/pdmat.jl#line=98) [inlined]
 [5] pointwise_loglikelihoods!(log_like::SubArray{Float64, 1, OffsetMatrix{Float64, Matrix{Float64}}, Tuple{Int64, Base.Slice{OffsetArrays.IdOffsetRange{Int64, Base.OneTo{Int64}}}}, true}, y::OffsetVector{Float64, Vector{Float64}}, dist::MvNormal{Float64, PDMat{Float64, Matrix{Float64}}, OffsetVector{Float64, Vector{Float64}}})
   @ PosteriorStats [~/Desktop/work/PosteriorStats.jl/src/pointwise_loglikelihoods.jl:23](http://localhost:8888/home/mbonici/Desktop/work/PosteriorStats.jl/src/pointwise_loglikelihoods.jl#line=22)
 [6] pointwise_loglikelihoods(y::OffsetVector{Float64, Vector{Float64}}, dists::OffsetVector{MvNormal{Float64, PDMat{Float64, Matrix{Float64}}, OffsetVector{Float64, Vector{Float64}}}, Vector{MvNormal{Float64, PDMat{Float64, Matrix{Float64}}, OffsetVector{Float64, Vector{Float64}}}}})
   @ PosteriorStats [~/Desktop/work/PosteriorStats.jl/src/pointwise_loglikelihoods.jl:11](http://localhost:8888/home/mbonici/Desktop/work/PosteriorStats.jl/src/pointwise_loglikelihoods.jl#line=10)
 [7] top-level scope
   @ In[157]:16
```
 
If I change `Σ = OffsetArray(Matrix(I, 3, 3), 0:2, 0:2)` as well I get
```julia
DimensionMismatch: axes must agree, got (Base.OneTo(3), Base.OneTo(3)) and (OffsetArrays.IdOffsetRange(values=0:2, indices=0:2), OffsetArrays.IdOffsetRange(values=0:2, indices=0:2))

Stacktrace:
 [1] _checkaxs(axd::Tuple{Base.OneTo{Int64}, Base.OneTo{Int64}}, axs::Tuple{OffsetArrays.IdOffsetRange{Int64, Base.OneTo{Int64}}, OffsetArrays.IdOffsetRange{Int64, Base.OneTo{Int64}}})
   @ Base ./abstractarray.jl:1163
 [2] copyto_axcheck!
   @ ./abstractarray.jl:1166 [inlined]
 [3] Array
   @ ./array.jl:626 [inlined]
 [4] convert
   @ ./array.jl:618 [inlined]
 [5] PDMat{Float64, Matrix{Float64}}(m::OffsetMatrix{Bool, Matrix{Bool}}, c::Cholesky{Float64, Matrix{Float64}})
   @ PDMats [~/.julia/packages/PDMats/cb4LS/src/pdmat.jl:15](http://localhost:8888/home/mbonici/.julia/packages/PDMats/cb4LS/src/pdmat.jl#line=14)
 [6] PDMat
   @ [~/.julia/packages/PDMats/cb4LS/src/pdmat.jl:22](http://localhost:8888/home/mbonici/.julia/packages/PDMats/cb4LS/src/pdmat.jl#line=21) [inlined]
 [7] PDMat
   @ [~/.julia/packages/PDMats/cb4LS/src/pdmat.jl:40](http://localhost:8888/home/mbonici/.julia/packages/PDMats/cb4LS/src/pdmat.jl#line=39) [inlined]
 [8] MvNormal(μ::OffsetVector{Float64, Vector{Float64}}, Σ::OffsetMatrix{Bool, Matrix{Bool}})
   @ Distributions [~/.julia/packages/Distributions/YQSrn/src/multivariate/mvnormal.jl:197](http://localhost:8888/home/mbonici/.julia/packages/Distributions/YQSrn/src/multivariate/mvnormal.jl#line=196)
 [9] top-level scope
   @ In[163]:8
```

I am very confused, especially because I am not even sure I am doing what you are asking me to do :sweat_smile: , I think I will need some guidance @sethaxen , sorry for that :(

sethaxen commented:
> I have some troubles with the `OffsetArray`s stuff.
> 
> I am very confused, especially because I am not even sure I am doing what you are asking me to do 😅 , I think I will need some guidance @sethaxen , sorry for that :(

No worries, this is definitely weird. So it seems here we have 2 issues:

1. OffsetArrays is enforcing that when two arrays are subtracted, they must have the same axes. That strict check makes sense for OffsetArrays, but not for the case we really care about: dimensional arrays
2. OffsetArrays doesn't support LinearAlgebra (specifically `\` with `Cholesky` on an `OffsetArray`). That's a no-go here.

So instead, let's test with DimensionalData directly, which probably has the most complete of all of the dimensional array implementations. This works for me:
```julia
julia> using DimensionalData

julia> y = DimArray(randn(3), Y(1:3));

julia> dists = fill(d, (Dim{:draw}(1:100), Dim{:chain}(1:4)));  # a show method errors here if we display this, but that's not a problem

julia> pll = PosteriorStats.pointwise_loglikelihoods(y, dists)
```
Then you can just check:
```julia
@test pll isa DimArray
@test dims(pll) == (dims(dists)..., dims(y)...)
```

sethaxen commented:
For now, doctest failures on Documenter and CI failures on `pre` are expected (due to a circular dependency issue with tests/docs and the InferenceObjects/PosteriorStats extension I haven't found a workaround for yet), so don't worry about those.

marcobonici commented:
Ok!
I have implemented your suggestion for the `DimensionalData` test. I have included all relevant dependencies and included some tests. The tests ranges from simple checks (when  Dense covariance matrix is used for the MvNormal), internal consistency (if the MVNormalCanon is constructed with the inverse of the Covariance provided to MVNormal, the results should be the same), to checks that the promotion when using `Float32` is correct.

Let me know if there is anything important I am missing.

sethaxen commented:
@marcobonici I refactored the tests and make a couple of code tweaks to support older versions of Distributions/PDMats. While I was at it, I added the docstring and docs entry. Would you like to take a look and let me know if you think anything could be clearer?

marcobonici commented on src/pointwise_loglikelihoods.jl:
```diff
@@ -0,0 +1,102 @@
+"""
+    pointwise_loglikelihoods
+
+Compute pointwise conditional log-likelihoods for use in ELPD-based model comparison.
+
```
Should we say this also for LOO validation in general, as LOO-PIT is not about model comparison?

marcobonici commented:
> @marcobonici I refactored the tests and make a couple of code tweaks to support older versions of Distributions/PDMats. While I was at it, I added the docstring and docs entry. Would you like to take a look and let me know if you think anything could be clearer?

Hi @sethaxen , happy with what you wrote (and I actually feel bad because of the higher quality of your tests).
I just added a minor comment in the docstring, but for me everything is clear :)
